File: Assets/Code/Logic/Shop/Model/ShopGoodsData.cs
using System.Collections.Generic;
using Common.Util;
using Logic.Game.Model;
using Logic.Item.Model;
using Logic.Hero.Model;

namespace Logic.Shop.Model
{
	public class ShopGoodsData
	{
		private static Dictionary<int, ShopGoodsData> _shopGoodsDataDictionary;
		public static Dictionary<int, ShopGoodsData> GetShopGoodsData ()
		{
			if (_shopGoodsDataDictionary == null)
			{
				_shopGoodsDataDictionary = CSVUtil.Parse<int, ShopGoodsData>("config/csv/shop_goods", "id");
			}
			return _shopGoodsDataDictionary;
		}

		[CSVElement("id")]
		public int id;

		[CSVElement("sheet_num")]
		public int shopID;

		[CSVElement("buy_type")]
		public int buyType;

		public GameResData goodsGameResData;
		[CSVElement("item")]
		public string item
		{
			set
			{
				goodsGameResData = new GameResData(value);
			}
		}

		[CSVElement("name")]
		public string overrideName;
		public string name
		{
			get
			{
				string realName = overrideName;
				if (string.IsNullOrEmpty(realName))
				{
					if (goodsGameResData.type == Logic.Enums.BaseResType.Item)
					{
						realName = ItemData.GetItemDataByID(goodsGameResData.id).name;
					}
					else if (goodsGameResData.type == Logic.Enums.BaseResType.Hero)
					{
						realName = HeroData.GetHeroDataByID(goodsGameResData.id).name;
					}
				}
				return realName;
			}
		}

		[CSVElement("des")]
		public string overrideDescription;
		public string description
		{
			get
			{
				string realDescription = overrideDescription;
				if (string.IsNullOrEmpty(realDescription))
				{
					if (goodsGameResData.type == Logic.Enums.BaseResType.Item)
					{
						realDescription = ItemData.GetItemDataByID(goodsGameResData.id).des;
					}
					else if (goodsGameResData.type == Logic.Enums.BaseResType.Hero)
					{
						realDescription = HeroData.GetHeroDataByID(goodsGameResData.id).description;
					}
				}
				return realDescription;
			}
		}

		[CSVElement("pic")]
		public string pic;

		public GameResData costGameResData;
		[CSVElement("cost")]
		public string cost
		{
			set
			{
				costGameResData = new GameResData(value);
			}
		}

		[CSVElement("item_num")]
		public int itemNum;
	}
}

File: Assets/ToLua/Lua/user/ui/shop/model/data/shop_goods_data.lua
local t = {}
t.data = {}
local item = {}
item.__index = item

function item.New(table)
	local o = {}
	setmetatable(o,item)
	o.id = tonumber(table.id)
	o.sheet_num = tonumber(table.sheet_num)
	o.buy_type = tonumber(table.buy_type)
	o.item = table.item
	o.name = table.name
	o.des = table.des
	o.pic = table.pic
	o.cost = table.cost
	o.item_num = tonumber(table.item_num)
  o.discount = tonumber(table.discount)
	return o
end

function t.GetDataById(id)
	return t.data[id]
end

local function Start()
	local origin = dofile('shop_goods')
	origin.ForEach(function(id,table)
		local newItem = item.New(table)
		t.data[id] = newItem
	end)
end

Start()
return t